crates/uv-workspace/src/pyproject.rs

```diff
@@ -407,6 +409,90 @@ impl Deref for SerdePattern {
     }
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[serde(rename_all = "kebab-case", try_from = "SourcesWire")]
+pub struct Sources(Vec<Source>);
+
+impl Sources {
+    /// Return an [`Iterator`] over the sources.
+    pub fn iter(&self) -> impl Iterator<Item = &Source> {
+        self.0.iter()
+    }
+}
+
+impl IntoIterator for Sources {
+    type Item = Source;
+    type IntoIter = std::vec::IntoIter<Source>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case", untagged)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+enum SourcesWire {
+    One(Source),
+    Many(Vec<Source>),
+}
+
+impl TryFrom<SourcesWire> for Sources {
+    type Error = SourceError;
+
+    fn try_from(wire: SourcesWire) -> Result<Self, Self::Error> {
+        match wire {
+            SourcesWire::One(source) => Ok(Self(vec![source])),
+            SourcesWire::Many(sources) => {
+                // Ensure that the markers are disjoint.
+                for (lhs, rhs) in sources
+                    .iter()
+                    .map(Source::marker)
+                    .zip(sources.iter().skip(1).map(Source::marker))
+                {
+                    if !lhs.is_disjoint(&rhs) {
+                        let mut hint = lhs.negate();
+                        hint.and(rhs.clone());
+
+                        let lhs = lhs
+                            .contents()
+                            .map(|contents| contents.to_string())
+                            .unwrap_or("true".to_string());
+                        let rhs = rhs
+                            .contents()
+                            .map(|contents| contents.to_string())
+                            .unwrap_or("true".to_string());
+                        let hint = hint
+                            .contents()
+                            .map(|contents| contents.to_string())
+                            .unwrap_or("true".to_string());
+
+                        return Err(SourceError::OverlappingMarkers(lhs, rhs, hint));
+                    }
+                }
+
+                // Ensure that there is at least one source.
+                if sources.is_empty() {
+                    return Err(SourceError::EmptySources);
+                }
+
+                // Ensure that there is at most one registry source.
+                if sources
+                    .iter()
+                    .filter(|source| matches!(source, Source::Registry { .. }))
+                    .nth(1)
+                    .is_some()
+                {
+                    return Err(SourceError::MultipleIndexes);
+                }
+
+                Ok(Self(sources))
+            }
+        }
+    }
+}
```

Comment:
@BurntSushi -- I'd love your eyes on this file in particular. Should I be enforcing these at deserialization time? I could easily enforce them when we actually _use_ the lowered requirement. I'm not certain what's best.